### ISSUE TITLE
SPIKE: Gimme a TreeGrid

### ISSUE THREAD

**Issue by pleku** [vaadin/flow#2822]
Take a look at the hierarchical features of `vaadin-grid` and try to make sense on how to integrate it to the `HierarchyDataProvider` and `HierarchyDataCommunicator` that were used in FW8.

**Comment by gilberto-torrezan** [comment]
Spike of the TreeGrid in Flow can be found in the branch https://github.com/vaadin/flow/tree/gilberto/2822_tree_grid_spike

* It is not a fully functional implementation *

A first level of the TreeGrid is rendered successfully, as shown in the screenshot:

![treegrid](https://user-images.githubusercontent.com/7464735/32553886-c24d1dde-c4a0-11e7-8e6f-f5f8fc1e88ee.png)

... but the problem is to support the remaining levels.

![treegrid_bug](https://user-images.githubusercontent.com/7464735/32553951-f60bda84-c4a0-11e7-8d30-ab3fe9e52092.png)

(This happens when clicking to open the first directory in the model. Note that the model repeats itself endlessly)

In the spike, the API from FW8 was ported to Flow, including the `HierarchicalDataCommunicator`, `HierarchicalDataProvider`, `HierarchyMapper` and the `TreeGrid` itself. Most of the ported classes could be compiled with no or small changes, but some, like the data communicator, needs design changes to properly be ported without ugly workarounds.

For instance - the `HierarchyMapper` is a `DataGenerator`, but the `Grid` expects internally to have a `GridDataGenerator`, which in turn accepts other generators inside it for a chain of generators. The `HierarchicalDataCommunicator` has no relation to the `Grid`, so it doesn't know about that extension mechanism, and things don't go nicely when it tries to force the `HierarchyMapper` as the only generator. This is fixed by a hack in the spike, but it should be properly redesigned in a proper implementation.

At the client side, the main problem is with the internal cache at the `gridConnector.js` file. That file is the "glue code" between the client and the server-sides  of `Grid` to make all the communications work.

When the webcomponent asks for new items to show, via its `dataprovider` function, a server-side method is called (`grid.$server.setRequestedRange`), and a callback to the webcomponent is stored. The server computes the items to be sent, and calls a function in the connector (`grid.$connector.set`), setting the items in an internal cache. And then, with the items in the cache, the stored callback is finally executed (`grid.$connector.confirm`), using the stored items to set them directly into the grid's cache, and revalidates the view.

The problem is that this intermediate cache is a flat one - it doesn't consider any hierarchy. And also there are some calculations about the row index of a specific range of items, which is also broken when considering hierarchical data.

So, one way of implementing the `TreeGrid` support at the connector is to create a complete different script to deal with hierarchical data - that wouldn't break any compatibility with the regular Grid, since it would be a different script. After the script is created, the common parts could be extracted to parent script, or something else that both scripts could use.

(Note - in the spike, the server method `grid.$server.setRequestedRange` was replaced by a `grid.$server.setRequestedRangeAndParentKey`, which also sends the parent key with was expanded in the client. That's only supported in the `TreeGrid`).

Also this spike didn't consider any server-side collapse/expand of items, nor collapse/expand events.

So, in summary:

Server-side:
- Small changes in most classes
- Needs redesign of the interaction between `HierarchicalDataCommunicator`, `HierarchyMapper` and `TreeGrid`
 - Most features work out-of-the-box

Client-side:
 - Refactor of the intermediate cache to support hierarchical data
 - Refactor of the calculations involving row numbers and pages
 - Different calls to the server when sending parent keys together with ranges
